src/charm.py

```diff
@@ -83,6 +103,20 @@ def __init__(self, *args):
             self._on_loki_push_api_endpoint_departed,
         )
 
+    def update_metrics_rules(self, _):
+        """Copy alert rules from relations and save them to disk."""
```

Comment:
Could you please add some more context?
I am having difficulty understanding what/why this function is for.

Reply:
It deletes all existing rules, gets new ones from the relation, writes them out to files, and triggers a reload.  It could probably be named `regenerate_metrics_rules` for accuracy.

The actual method is kind of tortured, though. It would be nice to see exception handling for `rmtree()`, even if it's unlikely.

Take it or leave it, but I'd rewrite the block trying to use "native" Pathlib stuff for consistency. It ends up a lot shorter, which isn't that important, but I'd find myself spending less time wondering "are we going to use this or that variable later".

```python
for topology_identifier, rule in rules.items():
  file = Path(self._metrics_rules_path, "juju_{}_rules".format(topology_identifier))
  file.write(yaml.dump(rule))
  logger.debug(...)
```

Reply:
I updated the method to use pathlib.

Reply:
Thanks @rbarry82.
@dstathis I think it would be nice to include in the docstring an answer to the "why" as well.

Reply:
I kind of disagree there. If we think of docstrings as primarily "references seen by developers either when editing the code or when importing it", the "why" is less important here.

From a `README` point of view for "how does this charm work?", sure. It's not necessarily important to know "why" they need to be written to disk to understand what the method is supposed to do, though, only that it's required.

Reply:
> I kind of disagree there. If we think of docstrings as primarily "references seen by developers either when editing the code or when importing it", the "why" is less important here.
> 
> From a `README` point of view for "how does this charm work?", sure. It's not necessarily important to know "why" they need to be written to disk to understand what the method is supposed to do, though, only that it's required.

The why might also change, or expand, during the lifetime of the piece of code